src/openclsim/critical_path/base_cp.py

```diff
@@ -66,9 +70,197 @@ def _make_recorded_activities_df(self):
+        """
+        # keep the df chronological
+        df_log = df_log.sort_values(by=["Timestamp", "ActivityState"])
+        df_log = df_log.reset_index()
+
+        # make a list of indexes to handle
+        to_handle = list(range(0, len(df_log)))
+
+        # init the output
+        recorded_activities_df = pd.DataFrame()
+
+        # loop exit
+        safety_valve = 0
+        while (len(to_handle) > 0) and (safety_valve < len(df_log)):
+            # update the safety valve
+            safety_valve += 1
+
+            # select a log-row to inspect
+            idx_start = to_handle[0]
+            row_current = df_log.loc[idx_start, :]
+
+            # check for a start event
+            if row_current.loc["ActivityState"] not in ["START", "WAIT_START"]:
+                raise ValueError(
+                    f"Unexpected starting state {row_current.loc['ActivityState']}"
+                    f" for idx {idx_start}, so skipping this."
+                )
+
+            # see what stop events could belong to this start event
+            bool_candidates = (
+                (df_log.loc[:, "ActivityID"] == row_current.loc["ActivityID"])
+                & (
+                    df_log.loc[:, "SimulationObject"]
+                    == row_current.loc["SimulationObject"]
+                )
+                & (df_log.loc[:, "ActivityState"].isin(["STOP", "WAIT_STOP"]))
+            )
+            idx_candidates = list(bool_candidates.index[bool_candidates])
+            # select the first end event after the start event
+            idx_end = [
+                idx_end
+                for idx_end in idx_candidates
+                if idx_end > idx_start and idx_end in to_handle
+            ][0]
+
+            # now remove idx start and end from handle
+            to_handle.remove(idx_start)
+            to_handle.remove(idx_end)
+
+            # and place in new dataframe
+            recorded_activities_df = pd.concat(
+                [
+                    recorded_activities_df,
+                    pd.DataFrame(
+                        {
+                            "Activity": row_current.loc["Activity"],
+                            "ActivityID": row_current.loc["ActivityID"],
+                            "SimulationObject": row_current.loc["SimulationObject"],
+                            "start_time": row_current.loc["Timestamp"],
+                            "state": "WAITING"
+                            if "WAIT" in row_current.loc["ActivityState"]
+                            else "ACTIVE",
+                            "duration": df_log.loc[idx_end, "Timestamp"]
+                            - row_current.loc["Timestamp"],
+                            "end_time": df_log.loc[idx_end, "Timestamp"],
+                        },
+                        index=[0],
+                    ),
+                ],
+                ignore_index=True,
+                sort=False,
+            )
```

Comment:
This part is once again very inefficient. I think the following does the same:

```python
df_log = all_recorded_events

df_log = df_log.sort_values(by=["ActivityID","Timestamp", "ActivityState"]).reset_index()

df_starts = (
    df_log[df_log.ActivityState.isin(["START", "WAIT_START"])]
    .filter(
        ["ActivityID","Activity","SimulationObject", "Timestamp", 'ActivityState']
    )
    .rename(columns={"Timestamp": "start_time"})
    .set_index("ActivityID")
)
df_stops = (
    df_log[df_log.ActivityState.isin(["STOP", "WAIT_STOP"])]
    .filter(["ActivityID", "Timestamp"])
    .rename(columns={"Timestamp": "end_time"})
    .set_index("ActivityID")
)
recorded_activities_df = pd.concat([df_starts, df_stops], axis=1).sort_values(by=["start_time"]).reset_index()

recorded_activities_df['duration'] = recorded_activities_df.end_time - recorded_activities_df.start_time
recorded_activities_df['state'] = recorded_activities_df.apply(lambda x: "WAITING" if "WAIT" in x["ActivityState"] else "ACTIVE", axis=1)

recorded_activities_df = recorded_activities_df.drop(columns=['ActivityState'])
```

Reply:
YES this is much nicer :)
Great change

Reply:
so implemented and resolved